### PR TITLE
[Experimental] Add new server endpoint pruning algorithm

### PR DESCRIPTION
**Pull request by schmittjoseph**
###### Summary

This PR adds a new, experimental, pruning algorithm for server endpoints. It is explicitly opt-in via an undocumented environment variable (`DotnetMonitor_Experimental_ServerEndpointPruningAlgorithmV2`). As we try this new mode out and build confidence in it, we can replace the current algorithm with it.

New Algorithm:
- When retrieving a list of active endpoints (`IServerEndpointTracker.GetEndpointInfoAsync`) return the last set of validated endpoints. It does **not** re-check the endpoints.
- Periodically check the state of endpoints. If one becomes unresponsive, it now has a 1 minute grace period to become active again before being pruned. While the endpoint is unresponsive but in this grace period, we still report the endpoint as active.

NOTE: I haven't updated any integration tests to use this new pruning algorithm as it's opt-in, undocumented.

### CODE REVIEW

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/EndpointInfo/ServerEndpointTrackerV2Tests.cs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+            // Act
+            await tracker.PruneEndpointsAsync(cts.Token);
+
+            // Assert
+            IEnumerable<IEndpointInfo> activeEndpoints = await tracker.GetEndpointInfoAsync(cts.Token);
+            Assert.Empty(activeEndpoints);
+
+            Assert.NotNull(endpointRemovedArgs);
+            Assert.Equal(ServerEndpointState.Error, endpointRemovedArgs.State);
+            Assert.Equal(endpointRemovedArgs.Endpoint, endpoint);
```
Comment (jander-msft) [inline src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/EndpointInfo/ServerEndpointTrackerV2Tests.cs:88]: I think these parameters should be switched: `endpoint` is the expected value whereas `endpointRemovedArgs.Endpoint` is the tested value.

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/EndpointInfo/ServerEndpointTrackerV2Tests.cs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+            // Act
+            await tracker.PruneEndpointsAsync(cts.Token);
+
+            // Assert
+            IEnumerable<IEndpointInfo> activeEndpoints = await tracker.GetEndpointInfoAsync(cts.Token);
+            Assert.Empty(activeEndpoints);
+
+            Assert.NotNull(endpointRemovedArgs);
+            Assert.Equal(ServerEndpointState.Unresponsive, endpointRemovedArgs.State);
+            Assert.Equal(endpointRemovedArgs.Endpoint, endpoint);
```
Comment (jander-msft) [inline src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/EndpointInfo/ServerEndpointTrackerV2Tests.cs:165]: I think these parameters should be switched: `endpoint` is the expected value whereas `endpointRemovedArgs.Endpoint` is the tested value.

**Review by jander-msft — DISMISSED** [review]

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/EndpointInfo/ServerEndpointTrackerV2Tests.cs (hunk 1 above)
Comment (schmittjoseph) [reply src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/EndpointInfo/ServerEndpointTrackerV2Tests.cs:88]: Whoops, good catch

**Review by jander-msft — APPROVED** [review]

**Pull request merged**